cmd/controller/app/controller.go

```diff
@@ -309,22 +368,23 @@ func startLeaderElection(ctx context.Context, opts *options.ControllerOptions, l
 	if err != nil {
 		// We should never get here.
 		log.Error(err, "error creating leader election lock")
-		os.Exit(1)
-
+		return err
 	}
 
 	// Try and become the leader and start controller manager loops
-	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
-		Lock:          ml,
-		LeaseDuration: opts.LeaderElectionLeaseDuration,
-		RenewDeadline: opts.LeaderElectionRenewDeadline,
-		RetryPeriod:   opts.LeaderElectionRetryPeriod,
-		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: run,
```

Comment:
I quite like the way this used to be. A single function containing all the controller setup code, which would only be run after leader election....but looking at the leaderelection code, it launches this in an unmanaged goroutine....so perhaps that can't allow us to wait for everything to finish before exiting.

Reply:
Indeed, OnStartedLeading is run in a separate goroutine. The previous implementation runs half of the setup in the main thread and half of the setup in this OnStartedLeading goroutine. Also if there is no leader election, it would run everything in the main thread. Now I just use OnStartedLeading to signal to the main thread that it can continue with its execution. This is similar to how https://github.com/kubernetes-sigs/controller-runtime/blob/d5fde106f02c76d9011676b97f117cd7d5d087ee/pkg/manager/internal.go#L644-L686 does it.